## Part 1

In [88]:
def is_piece(n, n_r, n_c, symbols, max_r, max_c, star_symbol_position):
    l = len(str(n))
    r_start = n_r -1
    c_start = n_c -1
    r_end = n_r + 2
    c_end = n_c + l + 1
    #print("checking adjacent symbol for", n, n_r, n_c)
    res = False
    for r in range(r_start, r_end):
        for c in range(c_start, c_end):
            #print("r, c:", r, c)
            if (r, c) in symbols:
                #print(n, "adjacent with", r, c)
                res = True
            if (r, c) in star_symbol_position:
                star_symbol_position[(r,c)].append(n)
    return res

symbol_position = []
star_symbol_position = {}
numbers = []
max_c = 0
max_r = 0
with open('input.txt') as f:
    r = 0
    max_r = len(line)-1
    for line in f:
        #print(line.strip())
        max_c = len(line.strip())
        curr_number = ""
        start = -1
        for i, ch in enumerate(line.strip()):
            if not ch.isdigit() and ch != '.':
                symbol_position.append((r, i))
                if ch == '*':
                    star_symbol_position[(r, i)] = []
            
            if ch.isdigit():
                curr_number += ch
                if start == -1:
                    start = i
            else:
                if len(curr_number) > 0:
                    numbers.append((int(curr_number), r, start))
                    curr_number = ""
                    start = -1
        r+=1
        if len(curr_number) > 0:
            numbers.append((int(curr_number), r-1, start))
            curr_number = ""
            start = -1
#print(symbol_position)
#print(numbers)
#print("max_r, max_c", max_r, max_c)


sum1 = 0
for e in numbers:
    if is_piece(e[0], e[1], e[2], symbol_position, max_r, max_c, star_symbol_position):
        #print("is piece", e[0])
        sum1 += e[0]

print(sum1)


#Part 2
sum2 = 0
for k in star_symbol_position:
    if len(star_symbol_position[k]) == 2:
        sum2 += star_symbol_position[k][0] * star_symbol_position[k][1]
print(sum2)

519444
74528807


Part 1
1. create set (r,c) of position containing symbol
2. extract number
3. for every number, check surrounding

first try : 517752 too low -> bug discovered, number at right border were not "detected"

second try : 519215 still too low -> number from 1st bug had a r too high of 1

third try : 519444 correct

## Part 2
first try : 74528807 correct